core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/scheduling/WorkflowPipelinedRegions.scala

```diff
@@ -0,0 +1,135 @@
+      new BiconnectivityInspector[OperatorIdentity, DefaultEdge](getBlockingEdgesRemovedDAG())
+    biconnectivityInspector
+      .getConnectedComponents()
+      .forEach(component => {
+        val regionId = PipelinedRegionIdentity(workflow.getWorkflowId(), regionCount.toString())
+        val operatorArray = new ArrayBuffer[OperatorIdentity]()
+        component.vertexSet().forEach(opId => operatorArray.append(opId))
+        pipelinedRegionsDAG.addVertex(new PipelinedRegion(regionId, operatorArray.toArray))
+        regionCount += 1
+      })
+  }
+
+  private def getPipelinedRegionFromOperatorId(operatorId: OperatorIdentity): PipelinedRegion = {
+    pipelinedRegionsDAG
+      .vertexSet()
+      .forEach(region =>
+        if (region.getOperators().contains(operatorId)) {
+          return region
+        }
+      )
+    null
+  }
+
+  private def findDependenciesBetweenRegions(): Unit = {
+    val regionTerminalOperatorInOtherRegions =
+      new mutable.HashMap[PipelinedRegion, ArrayBuffer[OperatorIdentity]]()
+    val allOperatorIds = workflow.getAllOperatorIds
+    allOperatorIds.foreach(opId => {
+      // 1. Find dependencies between pipelined regions enforced by inputs of operators.
+      // e.g 2 phase hash join requires build input to come first.
+      val inputProcessingOrderForOp = workflow.getOperator(opId).getInputProcessingOrder()
+      if (inputProcessingOrderForOp != null && inputProcessingOrderForOp.length > 1) {
+        for (i <- 1 to inputProcessingOrderForOp.length - 1) {
+          val prevInOrder = getPipelinedRegionFromOperatorId(
+            toOperatorIdentity(inputProcessingOrderForOp(i - 1).from)
+          )
+          val nextInOrder =
+            getPipelinedRegionFromOperatorId(toOperatorIdentity(inputProcessingOrderForOp(i).from))
+
+          if (
+            prevInOrder.getId() != nextInOrder
+              .getId() && !pipelinedRegionsDAG.getDescendants(prevInOrder).contains(nextInOrder)
+          ) {
+            pipelinedRegionsDAG.addEdge(prevInOrder, nextInOrder)
+          }
+        }
+      }
+
+      // 2. Find dependencies due to blocking operators.
+      // e.g. The region before and after a sort operator has dependencies
+      val upstreamOps = workflow.getDirectUpstreamOperators(opId)
+      upstreamOps.foreach(upstreamOp => {
+        val linkFromUpstreamOp = LinkIdentity(
+          workflow.getOperator(upstreamOp).topology.layers.last.id,
+          workflow.getOperator(opId).topology.layers.head.id
+        )
+        if (workflow.getOperator(opId).isInputBlocking(linkFromUpstreamOp)) {
+          val prevInOrder = getPipelinedRegionFromOperatorId(upstreamOp)
+          val nextInOrder = getPipelinedRegionFromOperatorId(opId)
+          if (prevInOrder.getId() != nextInOrder.getId()) {
+            if (!pipelinedRegionsDAG.getDescendants(prevInOrder).contains(nextInOrder)) {
+              pipelinedRegionsDAG.addEdge(prevInOrder, nextInOrder)
+            }
+
+            if (
```

Comment:
ditto

Reply:
I dont think there is any exception required here.